### ISSUE TITLE
Provide control to determine if a project needs rebuild with --build flag

### ISSUE THREAD

**Issue by jasonaden** [microsoft/TypeScript#54563]
# Suggestion

## 🔍 Search Terms

composite build output invalidation

composite build output invalidation

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [X] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [X] This wouldn't change the runtime behavior of existing JavaScript code
* [X] This could be implemented without emitting different JS based on the types of the expressions
* [ ] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [X] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

A programatic way to determine if a project needs to be rebuilt, or if the output for any given input file is invalid.

## 📃 Motivating Example

The key motivating example here is CI where we would like to be able to rely on output from `--build` off `master` branch when running CI for a branch. Most projects (in our example, > 60 individual projects in the monorepo) don't actually need to be rebuilt, but because CI is using git to restore files, the timestamps are different and therefore invalidating all project builds.

## 💻 Use Cases

Primary use case is CI as mentioned above. A secondary use case would be for distributed developer cache so build outputs could be shared (using something like Nx Cloud).

**Comment by RyanCavanaugh** [comment]
> but because CI is using git to restore files, the timestamps are different and therefore invalidating all project builds.

If this is the case, how is this functionality supposed to even work (in a way that's faster than just redoing the build)? I don't understand how this would be possible.

**Comment by RyanCavanaugh** [comment]
@sheetalkamat any thoughts?

**Comment by sheetalkamat** [comment]
There are couple of ways you can do this today. But whether build is uptodate or not is determined based on multiple things including timestamps for input, output, presence of errors, change of compiler options etc. If you think its just timestamps thats tripping it, is build not updating just timestamps instead of rebuilding the whole thing. Whats the log from `tsc --b --v` and reason for building the project. 

Aside from that here are the ways to achieve this today.
1. Update timestamps for output files you think are uptodate?
2. Use `SolutionBuilder` api and use `getNextInvalidatedProject` to get the next project to build and if you think this shouldnt be updated, then you could update the timestamps for all the outputs for that project and use `invalidateProject` which is currently internal but we can make it public if thats what is needed to re-evaluate the projects.

**Comment by MorrisonCole** [comment]
I'm not sure if this is the right issue to continue the discussion - but I'm running into the same problem as #59152. 

I'm working with a large monorepo where build times are significant. What I want to do is cache the tsbuildinfo files in CI so that projects that are unchanged aren't rebuilt unnecessarily.

With git & GitHub actions at least, I'm having trouble since `tsc -b` uses last modified timestamps to check whether to rebuild, and git doesn't track/preserve those, whereas the GitHub cache action _does_.

Here's the reason I get for rebuild:

```
Project 'project/tsconfig.json' is out of date because output 'project/tsconfig.tsbuildinfo' is older than input 'project/tsconfig.json'
```

I tried updating the last modified of all the cache files after checkout / cache restore, but that doesn't work since the build is then skipped even if there are changes in the sources of the project. Doing the last modified timestamp update intelligently (e.g., checking the last updated timestamp in git and applying that) would be fairly slow and quite unwieldy I think, so I decided to leave it for now.

**Comment by alex-statsig** [comment]
Hey @RyanCavanaugh , https://github.com/microsoft/TypeScript/issues/59152 was closed as a dupe of this Issue. Could we re-open that issue? It sounds like @MorrisonCole also reported running into the same issue

**Comment by blake-newman** [comment]
@alex-statsig i've resolved this in our company. I'm planning to open source this as a github action in the near future (over the next month). 

I'll let you know when I've got it live 👍 

---

To resolve this we did the following:

- Default branch cache the outputted tsbuildinfo and types 
- PR apply timestamps of the cached files to the base commit
- PR apply timestamps of source files to the head commit 
- Run typecheck

The timestamp restoring part is important because this is how TS determines changes, as git doesn't store timestamp information on files (as otherwise touching files would cause changes). 

This flow allows us to run incrementally in CI, and only project (references) that have changed will be triggered.

**Comment by tmm1** [comment]
@blake-newman do you have anything to share here?

**Comment by MorrisonCole** [comment]
> Does `find . -type f -name "*.tsbuildinfo" -exec touch {} +` solves this?

Nope, because then the build cache's last modified timestamps match the source files last modified timestamps, and so nothing is built (i.e., TS thinks everything is up-to-date).

What @blake-newman shared seems like a decent approach to me but haven't had a chance to try it.

**Comment by blake-newman** [comment]
> @blake-newman do you have anything to share here?

Ill be publishing shortly, just need to get approval to make repo public 👍

**Comment by blake-newman** [comment]
Sorry for the very long delay 😢 . I completly forgot about this, i've now made public and here it is:

https://github.com/Attest/typescript-cache-action

**Comment by thelinuxlich** [comment]
@blake-newman should we commit tsbuildinfo files into the repo for this action?

**Comment by blake-newman** [comment]
@thelinuxlich no need to, as it uses github actions cache to save tsbuildinfo and outputed declaration files. 

In the readme there is a bit of a demo of a rough setup but i'll quickly run through

## Restore cache 

This step restores the cache, you can give it cache keys, and a base ref to restore from. In the below example this will restore the tsbuildinfo files and declaration files from master. 

The `cache-base-key` is used to provide some environment variables, so if the runner node version etc changes itll generate a new set of caches. 
The `cache-key` is used to specify the cache, ie it will restore closest match from both
The `base-ref` is specified to restore the timestamps of the cache to that of the base ref. 

So in this example it will restore the declaration and buildinfo files from cache, and assign the timestamp of last commit on main branch. You could use a tag for example. 

```
- name: Restore Typescript Cache
  uses: Attest/typescript-cache-action@main
  if: ${{ inputs.only-changed == 'true' }}
  with:
    cache-base-key: "${{ runner.os }}-$(node -v)-ts_cache"
    cache-key: ${{ github.sha }}
    base-ref: main
    files: |
      ./packages/*/types/**/* 
      ./packages/*.tsbuildinfo

```

## run the typecheck

As the cache is now restored to a stable point in time, typescript will now be able to detect files that have been changed and rebuild only project references that have been updated etc.

```
      - name: typecheck
        run: pnpm typecheck --verbose
```

## Save the cache

Save the cache when running on the base branch, so that PRs can pick up this cache and run in an incremental mode. 

```
- name: Save Typescript Cache
  if: ${{ github.workflow == 'main' }}
  uses: Attest/typescript-cache-action/save@main
  with:
    cache-base-key: "${{ runner.os }}-$(node -v)-ts_cache"
    cache-key: ${{ github.sha }}
    files: |
      ./packages/*/types/**/* 
      ./packages/*.tsbuildinfo
```

---
I'll look to add some demo examples in the repository.

**Comment by thelinuxlich** [comment]
I see, thanks for the explanation!

**Comment by blake-newman** [comment]
You can look at the two actions:

restoring: https://github.com/Attest/typescript-cache-action/blob/main/action.yml
saving: https://github.com/Attest/typescript-cache-action/blob/main/save/action.yml

They are relatively simple as they just utilise other actions to do the above flows. So if you come across any issues, you can either fork the repo or recreate them inline as you see fit to resolve for your processes. 

Best of luck with getting incremental builds in CI. Feel free to reach out and i can try support best i can.
